pkg/download/downloader.go

```diff
@@ -162,20 +163,44 @@ func GetInsecure(uri, dir string, fetcher Fetcher) error {
 	return extractArchive(name, dir, body, size)
 }
 
-func extractArchive(filename, dst string, r io.ReaderAt, size int64) error {
+func extractContentType(at io.ReaderAt) (string, error) {
+	buf := make([]byte, 512)
+	n, err := at.ReadAt(buf, 0)
+	if err != nil && err != io.EOF {
+		return "", errors.Wrap(err, "failed to read first 512 bytes")
+	}
+	if n < 512 {
+		glog.V(5).Infof("Did only read %d of 512 bytes to determine the file type", n)
+	}
+	return http.DetectContentType(buf[:n]), nil
+}
+
+type extractor func(targetDir string, read io.ReaderAt, size int64) error
+
+var defaultExtractors = map[string]extractor{
+	"application/zip":      extractZIP,
+	"application/tar+gzip": extractTARGZ,
+	"application/x-gzip":   extractTARGZ,
+}
+
+func extractArchive(filename, dst string, at io.ReaderAt, size int64) error {
+	// TODO: Keep the filename for later direct download
+
 	// TODO(ahmetb) This package is not architected well, this method should not
 	// be receiving this many args. Primary problem is at GetInsecure and
 	// GetWithSha256 methods that embed extraction in them, which is orthogonal.
 
 	// TODO(ahmetb) write tests with this by mocking extractZIP function into a
 	// zipExtractor variable and check its execution.
-
-	if strings.HasSuffix(filename, ".zip") {
-		glog.V(4).Infof("detected .zip file")
-		return extractZIP(dst, r, size)
-	} else if strings.HasSuffix(filename, ".tar.gz") {
-		glog.V(4).Infof("detected .tar.gz file")
-		return extractTARGZ(dst, io.NewSectionReader(r, 0, size))
+	t, err := extractContentType(at)
+	if err != nil {
+		return errors.Wrap(err, "failed to determine content type")
+	}
+	glog.V(4).Infof("detected %q file type", t)
+	exf, ok := defaultExtractors[t]
+	if !ok {
+		return errors.Errorf("cannot infer a supported archive type mime: (%s)", t)
 	}
-	return errors.Errorf("cannot infer a supported archive type from filename in the url (%q)", filename)
+	return errors.Wrap(exf(dst, at, size), "failed to extract file")
```

Comment:
I'm a bit confused about this `io.ReaderAt` here.

- This method gets `at io.ReaderAt` which this method doesn't know where it starts.
- `at` is passed to `detectContentType`
- `detectContentType` does `at.ReadAt(buf, 0)` which I'm not sure is a good idea. We could just pass this method an `io.Reader` or `[]byte` and have a cleaner design.
- `detectContentType` caused the current `Read()` cursor to advance (?) as a side-effect.
- we pass the same `at` to `exf(dst,at,size)`.
- now the extractZIP or extractTARGZ function has received an already-advanced reader. so they **have to** use a section reader (they previously weren't) to set the cursor back to zero.

Instead, I wish we could just 
- read first 512 bytes into []byte
- pass []byte to detectContentType()
- reset the `at` to cursor position = 0.
- pass `at` down so the downstream extractors (extractZIP etc) don't need to explicitly say "start reading from 0" and they can just assume it's already at zero (=unconsumed reader).